README
[[!meta title="cdist - usable configuration management"]]

     
                 ..          .       .x+=:.        s   
               dF           @88>    z`    ^%      :8  
              '88bu.        %8P        .   <k    .88 
          .   '*88888bu      .       .@8Ned8"   :888ooo
     .udR88N    ^"*8888N   .@88u   .@^%8888"  -*8888888
    <888'888k  beWE "888L ''888E` x88:  `)8b.   8888
    9888 'Y"   888E  888E   888E  8888N=*8888   8888
    9888       888E  888E   888E   %8"    R88   8888
    9888       888E  888F   888E    @8Wou 9%   .8888Lu=
    ?8888u../ .888N..888    888&  .888888P`    ^%888*
     "8888P'   `"888*""     R888" `   ^"F        'Y" 
       "P'        ""         ""  
     

[[!toc levels=3]]

## Introduction

cdist is a usable configuration management system. It adheres to
the KISS principle and is being used in small up to enterprise grade
environments.
cdist is an alternative to other configuration management systems like
[cfengine](http://www.cfengine.org/),
[bcfg2](http://trac.mcs.anl.gov/projects/bcfg2),
[chef](http://wiki.opscode.com/display/chef/)
and [puppet](http://www.puppetlabs.com/).
But cdist ticks differently, here is the feature set that makes it unique: 

[[!table  data="""
Keywords | Description
Simplicity | There is only one type to extend cdist called ***type***
Design | Type and core cleanly seperated
Design | Sticks completly to the KISS (keep it simple and stupid) paradigma
Design | Meaningful error messages - do not lose time debugging error messages
Design | Consistency in behaviour, naming and documentation
Design | No surprise factor: Only do what is obviously clear, no magic
Design | Define target state, do not focus on methods or scripts
Design | Push architecture: Instantly apply your changes
Small core | cdist's core is very small - less code, less bugs
Fast development | Focus on straightforwardness of type creation is a main development objective
Fast development | Batteries included: A lot of requirements can be solved using standard types
Modern Programming Language | cdist is written in Python
Requirements, Scalability | No central server needed, cdist operates in push mode and can be run from any computer
Requirements, Scalability, Upgrade | cdist only needs to be updated on the master, not on the target hosts
Requirements, Security | Uses well-know [SSH](http://www.openssh.com/) as transport protocol
Requirements, Simplicity | Requires only shell and SSH server on the target
UNIX | Reuse of existing tools like cat, find, mv, ...
UNIX, familar environment, documentation | Is available as manpages and HTML
UNIX, simplicity, familar environment | cdist is configured in POSIX shell
"""]]

### Documentation

The cdist documentation is included as manpages in the distribution.  
You can browse the documentation online as well:

 * [latest version](man/latest)
 * [all versions (>= 2.0.4)](man)

### OS support

cdist was tested or is know to run on at least

 * [Archlinux](http://www.archlinux.org/)
 * [Debian](http://www.debian.org/)
 * [CentOS](http://www.centos.org/)
 * [Fedora](http://fedoraproject.org/)
 * [Gentoo](http://www.gentoo.org/)
 * [Mac OS X](http://www.apple.com/macosx/)
 * [OpenBSD](http://www.openbsd.org)
 * [Redhat](http://www.redhat.com/)
 * [Ubuntu](http://www.ubuntu.com/)
 * [XenServer](http://www.citrix.com/xenserver/)


## Requirements

### Server

 * A posix like shell
 * Python (>= 3.2 required)
 * SSH client
 * Asciidoc (for building the manpages)

### Client ("target host")

 * A posix like shell
 * SSH server


## Installation

### Preparation

Ensure you have Python 3.2 installed on the machine you use to **deploy to the targets**
(the ***source host***).

#### Archlinux

Archlinux already has python >= 3.2, so you only need to do:

    pacman -S python

#### CentOS

See the "From source" section

#### Debian

For Debian >= wheezy:

    aptitude install python3

On squeeze you can add following line in **/etc/apt/sources.list**

  deb http://ftp.debian.org/debian wheezy main

And add pinning entry in **/etc/apt/preferences.d/wheezy**:

  Package: *
  Pin: release n=wheezy
  Pin-Priority: 1

Please be aware that both **openssh-server** and **openssh-client** might be
removed on **python3.2** installation. You surely want to reinstall them:

  apt-get install -t wheezy openssh-server openssh-client

For older Debian versions, installing python 3.2 manually is required.


#### Fedora

For Fedora >= 15:

    yum install python3

#### FreeBSD

For the port:

    cd /usr/ports/lang/python32/ && make install clean

For the package:

    pkg_add -r python32

#### Gentoo

Gentoo only provides python 3.2 in testing packages (http://www.gentoo.org/doc/en/handbook/handbook-x86.xml?part=3&chap=3).
If you want to ensure nothing breaks you must set back the python version to what was default before.

    emerge -av =python-3.2.2 --autounmask-write
    emerge -av =python-3.2.2
    eselect python list
    eselect python list set python3.2

#### Max OS X

You can choose between Homebrew and Macports, either way works:

[Homebrew](http://mxcl.github.com/homebrew/) variant:

    brew install python3

[Macports](http://www.macports.org/install.php) variant:

    port install python32
    ln -s /opt/local/bin/python3.2 /opt/local/bin/python3

#### From Source

For those operating systems not yet support Python 3.2:

    pyversion=3.2.3
    wget http://www.python.org/ftp/python/$pyversion/Python-${pyversion}.tar.bz2
    tar xvfj Python-${pyversion}.tar.bz2
    cd Python-${pyversion}
    ./configure
    make
    sudo make install

This installs python 3.2 to /usr/local/bin. Ensure this directory is in
your PATH environment variable.


### Get cdist

You can clone cdist from git, which gives you the advantage of having
a version control in place for development of your own stuff as well.
To install cdist, execute the following commands:

    git clone git://git.schottelius.org/cdist
    cd cdist
    export PATH=$PATH:$(pwd -P)/bin

    # If you want the manpages
    ./build man
    export MANPATH=$MANPATH:$(pwd -P)/doc/man


### Available versions

There are at least the following branches available:

 * Development: master
 * 2.0: Python rewrite of cdist core [stable branch]

Old versions:

 * 1.7: Bugfixes, cleanups, new type and explorer rename
 * 1.6: New types, cleaned up \_\_package* types, internal cleanup
 * 1.5: Focus on object orientation instead of global stage orientation
 * 1.4: Support for redefiniton of objects (if equal)
 * 1.3: Support for local and remote code execution (current stable)
 * 1.2: Dependencies supported
 * 1.1: \_\_file to \_\_file, \_\_directory, \_\_link migration
 * 1.0: First official release

Other branches may be available for features or bugfixes, but they
may vanish at any point. To select a specific branch use

    # Generic code
    git checkout -b <name> origin/<name>
    
    # Stay on a specific version
    version=2.0
    git checkout -b $version origin/$version

### Mirrors

 * git://github.com/telmich/cdist.git ([github](https://github.com/telmich/cdist))
 * git://git.code.sf.net/p/cdist/code ([sourceforge](https://sourceforge.net/p/cdist/code))

## Update

To upgrade cdist in the current branch use

    git pull

    # Also update the manpages
    ./build man
    export MANPATH=$MANPATH:$(pwd -P)/doc/man

If you stay on a version branche (i.e. 1.0, 1.1., ...), nothing should break.
The master branch on the other hand is the development branch and may not be
working, break your setup or eat the tree in your garden.

### Upgrading from 1.7 to 2.0

* Ensure python (>= 3.2) is installed on the server
* Use "cdist config host" instead of "cdist-deploy-to host"
* Use "cdist config -p host1 host2" instead of "cdist-mass-deploy"
* Use "cdist banner" for fun
* Use **\_\_object_fq** instead of **\_\_self** in manifests

### Upgrading from 1.6 to 1.7

* If you used the global explorer **hardware_type**, you need to change
  your code to use **machine** instead.

### Upgrading from 1.5 to 1.6

* If you used **\_\_package_apt --preseed**, you need to use the new
  type **\_\_debconf_set_selections** instead.
* The **\_\_package** types accepted either --state deinstalled or
  --state uninstaaled. Starting with 1.6, it was made consistently
  to --state removed.

### Upgrading from 1.3 to 1.5

No incompatiblities.

### Upgrading from 1.2 to 1.3

Rename **gencode** of every type to **gencode-remote**.

### Upgrading from 1.1 to 1.2

No incompatiblities.

### Upgrading from 1.0 to 1.1

In 1.1 the type **\_\_file** was split into **\_\_directory**, **\_\_file** and
**\_\_link**. The parameter **--type** was removed from **\_\_file**. Thus you
need to replace **\_\_file** calls in your manifests:

 * Remove --type from all \_\_file calls
 * If type was symlink, use \_\_link and --type symbolic
 * If type was directory, use \_\_directory


## Support

### IRC

You can join the development ***IRC channel***
[#cstar on irc.freenode.net](irc://irc.freenode.org/#cstar).

### Mailing list

Bug reports, questions, patches, etc. should be send to the
[cdist mailing list](http://l.schottelius.org/mailman/listinfo/cdist).

### Linkedin

If you have an account
at [Linked in](http://www.linkedin.com/),
you can join the
[cdist group](http://www.linkedin.com/groups/cdist-configuration-management-3952797).

### Commercial support

You can request commercial support for cdist from
[my company](http://firma.schottelius.org/english/).

## Used by

If you're using cdist, feel free to send a report to the mailing list.
Interesting information are for instance

 * Which services do you manage?
 * How many machines do you manage?
 * What are the pros/cons you see in cdist?
 * General comments/critics

### Nico Schottelius, Systems Group ETH Zurich, local.ch and privately

Yes, I'm actually eating my own dogfood and currently managing

 * [plone](http://plone.org/) (cms)
 * [moinmoin](http://moinmo.in/) (wiki)
 * [apache](http://httpd.apache.org/) (webserver)
 * [kerberos (mit)](http://web.mit.edu/kerberos/) (authentication)
 * [nss-pam-ldapd](http://arthurdejong.org/nss-pam-ldapd/) (authentication)
 * [ircd-hybrid](http://www.ircd-hybrid.org/) (chat)
 * [stunnel](http://stunnel.mirt.net/) (SSL tunnel)
 * [mercurial-server](http://www.lshift.net/mercurial-server.html) (version control)
 * [xfce](http://www.xfce.org/) (lightweight desktop environment)
 * [slim](http://slim.berlios.de/) (graphical login manager for X11)

with cdist on more than **60** production machines of the
[Systems Group](http://www.systems.ethz.ch) at the
[ETH Zurich](http://www.ethz.ch) as well at home.

### Steven Armstrong, CBRG ETH Zurich

The CBRG is managing most of their compute clusters with cdist.

[[!tag cdist unix]]


doc/changelog
Changelog
---------

	* Changes are always commented with their author in (braces)
	* Exception: No braces means author == Nico Schottelius

2.0.15: 2012-11-02
	* Core: Make variable __object_name available in type explorers (Steven Armtrong)
	* New Type: __qemu_img
	* New Type: __line
	* New Type: __pf_apply (Jake Guffey)
	* New Type: __pf_ruleset (Jake Guffey)
	* Bugfix Type: __rvm: Make type work if rvm is already installed

2.0.14: 2012-09-07
	* Bugfix Type: __jail: Use correct variable (Jake Guffey)
	* Change Type: __jail: Parameter jailbase now optional (Jake Guffey)
	* Bugfix Type: __user: Use passwd database on FreeBSD (Jake Guffey)
	* Bugfix Type: __start_on_boot: Do not change parameters
	* Feature __user: Added support for BSDs (Sébastien Gross)
	* Feature __group: Added support for FreeBSD (Jake Guffey)
	* New Type: __package_zypper
	* Feature Types: Initial Support for SuSE Linux
	
2.0.13: 2012-06-05
	* Bugfix __ssh_authorized_key: Ensure it sets proper group (contradict)
	* Bugfix __addifnosuchline: Fixed quotes/interpolation bug ("a  b" became "a b")
	* New Explorer: interfaces (Sébastien Gross)
	* Feature core: Support reading from stdin in types (Steven Armstrong)
	* Feature core: Support multiple parameters for types (Steven Armstrong)
	* Feature __file: Support reading from stdin with - syntax (Steven Armstrong)

2.0.12: 2012-05-29
	* Core: Correctly raise error on Python < 3.2 (Steven Armtrong)
	* Core: Add support for --remote-exec and --remote-copy parameters
	* Documentation: Debian Squeeze hints (Sébastien Gross)

2.0.11: 2012-05-23
	* Fix insecure file/directory creation: Use umask 077

2.0.10: 2012-05-18
	* Cleanup __group: No getent gshadow in old Redhat, use groupmod -g
		(Matt Coddington)
	* Bugfix __package_yum: Missing cat
	* Bugfix __start_on_boot: Correctly use sed and quotes (Steven Armstrong)
	* Feature __file: Support for --state exists (Steven Armstrong)
	* Feature core: Make variable __manifest available to type manifests
	* Feature core: Correct parent dependency handling (Steven Armstrong)
	* Bugfix several types: Fix sed for FreeBSD (Istvan Beregszaszi)
	* New Type: __jail (Jake Guffey)
	* Change Type: __rvm*: --state present/absent not installed/remvoed (Evax Software)
	* Bugfix Type: __cron: Hide error output from crontab
	* Various smaller bugfixes (Chris Lamb)

2.0.9: 2012-03-12
	* Cleanup documentation: Fix environment variable list to be properly 
		displayed (Giel van Schijndel)
	* Cleanup documentation: Some minor corrections
	* New Type: __package_opkg (Giel van Schijndel)
	* New Type: __package_pkg_freebsd (Jake Guffey)
	* New Type: __mysql_database (Benedikt Koeppel)
	* Feature __package: Support for OpenWRT (Giel van Schijndel)
	* Feature __start_on_boot: Support for OpenWRT (Giel van Schijndel)
	* Feature __start_on_boot: Support for Amazon Linux (Matt Coddington)
	* New Example: Use rsync to backup files (Matt Coddington) 
	* Feature core: Exit non-zero, if configuration failed
	* Documentation: Describe how to do templating (Aurélien Bondis)

2.0.8: 2012-02-20
	* Bugfix core: Remove another nasty traceback when sending SIGINT (aka Ctrl-C)
	* Cleanup: Better hint to source of error
	* Cleanup: Do not output failing script, but path to script only
	* Cleanup: Remove support for __debug variable in manifests (Type != Core
		debugging)
	* Cleanup: Change __package_* to support absent/present (default state
		name now). The values removed/installed will be removed in cdist 2.1.
	* Cleanup: Change __process to support absent/present (default state
		name now). The values running/stopped will be removed in cdist 2.1.
	* Feature Core: Support boolean parameters (Steven Armstrong)

2.0.7: 2012-02-13
	* Bugfix __file: Use chmod after chown/chgrp (Matt Coddington)
	* Bugfix __user: Correct shadow field in explorer (Matt Coddington)
	* Bugfix __link: Properly handle existing links (Steven Armstrong)
	* Bugfix __key_value: More robust implementation (Steven Armstrong)
	* Bugfix __user: Fix for changing a user's group by name (Matt Coddington)
	* New Type: __package_pip
	* Bugfix/Cleanup: Correctly allow Object ID to start and end with /, but
		not contain //.

2.0.6: 2012-01-28
	* Bugfix __apt_ppa:
		Also remove the [ppa-name].list file, if empty. (Tim Kersten)
	* Bugfix __group:
		Referenced wrong variable name (Matt Coddington)
	* Feature __package_apt:
		Initial support for virtual packages (Evax Software)
	* Feature Core: Added new dependency resolver (Steven Armstrong)
	* Feature Explorer, __package_yum: Support Amazon Linux (Matt Coddington)
	* New Type: __rvm (Evax Software)
	* New Type: __rvm_gem (Evax Software)
	* New Type: __rvm_gemset (Evax Software)
	* New Type: __rvm_ruby (Evax Software)

2.0.5: 2012-01-18
	* Bugfix __key_value: Use correct delimiters
		(Steven Armstrong, Daniel Maher)
	* Cleanup: Explicitly require Python >= 3.2 (do not fail implicitly)
	* Documentation: (Re)write of the tutorial
	* Feature: __addifnosuchline supports matching on 
		regular expressions (Daniel Maher)
	* Feature: __directory, __file, __link:
		Add --state parameter (Steven Armstrong)
	* New Type: __package_luarocks (Christian G. Warden)
	* New Type: __cdistmarker (Daniel Maher)

2.0.4: 2011-11-18
	* Bugfix core: Remove traceback when sending SIGINT (aka Ctrl-C)
	* Bugfix core: Accept parameters with - in the name (Steven Armstrong)
	* Cleanup: __object_fq variable removed (never used)
	* Cleanup: Environment variable __self DEPRECATED, use __object_name instead
	* Cleanup: Environment variable __self scheduled for removal in cdist 2.1
	* Documentation: Many examples for use of __remote_* (Steven Armstrong)
	* Feature: Automatically require all used objects (Steven Armstrong)
	* New Type: __cron (Steven Armstrong)

2.0.3: 2011-10-18
	* Improved logging, added --verbose, by more quiet by default
	* Bugfix __user: Correct quoting (Steven Armstrong)
	* Bugfix __addifnosuchline: Falsely exited

oo.py
import os
import tempfile

# FIXME: change these to match your environment
os.environ['__cdist_base_dir'] = '/home/sar/vcs/cdist'
# FIXME: testing against the cache, change path
os.environ['__cdist_out_dir'] = '/home/sar/vcs/cdist/cache/sans-asteven-02.ethz.ch/out'


'''
cd /path/to/dir/with/this/file
ipython


In [1]: import oo

In [2]: t = oo.Type('__mkfs')

In [3]: t.
t.base_dir             t.is_install           t.list_type_names      t.name                 t.path                 
t.explorers            t.is_singleton         t.list_types           t.optional_parameters  t.required_parameters  

In [3]: t.path
Out[3]: '/home/sar/vcs/cdist/conf/type/__mkfs'

In [4]: t.required_parameters
Out[4]: ['type']

In [5]: t.optional_parameters
Out[5]: ['device', 'options', 'blocks']

In [6]: t.is
t.is_install    t.is_singleton  

In [6]: t.is_singleton
Out[6]: False

In [7]: o = oo.Object(t, 'dev/sda1')

In [8]: o.
o.base_dir           o.list_object_names  o.list_type_names    o.parameter          o.qualified_name     o.type
o.changed            o.list_objects       o.object_id          o.path               o.requirements       

In [8]: o.pa
o.parameter  o.path       

In [8]: o.path
Out[8]: '/home/sar/vcs/cdist/cache/sans-asteven-02.ethz.ch/out/object/__mkfs/dev/sda1/.cdist'

In [9]: o.changed
Out[9]: False

In [10]: o.changed = True

In [11]: # creates /home/sar/vcs/cdist/cache/sans-asteven-02.ethz.ch/out/object/__mkfs/dev/sda1/.cdist/changed

In [12]: o.changed
Out[12]: True

In [13]: o.changed = False

In [14]: # removes /home/sar/vcs/cdist/cache/sans-asteven-02.ethz.ch/out/object/__mkfs/dev/sda1/.cdist/changed

In [15]:


-----

In [1]: import oo

In [2]: oo.Type.
oo.Type.base_dir             oo.Type.is_install           oo.Type.list_type_names      oo.Type.mro                  oo.Type.path                 
oo.Type.explorers            oo.Type.is_singleton         oo.Type.list_types           oo.Type.optional_parameters  oo.Type.required_parameters  

In [2]: oo.Type.list
oo.Type.list_type_names  oo.Type.list_types       

In [2]: oo.Type.list_type_names()
Out[2]: 
['__addifnosuchline',
 '__apt_ppa',
 '__apt_update_index',
 '__autofs_map',
 '__autofs_master',
 '__debconf_set_selections',
 '__directory',
 '__file',
 '__group',
 '__issue',
 '__key_value',
 '__link',
 '__mkfs',
 '__motd',
 '__package',
 '__package_apt',
 '__package_pacman',
 '__package_pkg_openbsd',
 '__package_rubygem',
 '__package_yum',
 '__partition_msdos',
 '__partition_msdos_apply',
 '__postgres_database',
 '__postgres_role',
 '__process',
 '__removeline',
 '__ssh_authorized_key',
 '__timezone',
 '__user']

In [3]: list(oo.Type.list_types())
Out[3]: 
[<Type name=__addifnosuchline>,
 <Type name=__apt_ppa>,
 <Type name=__apt_update_index>,
 <Type name=__autofs_map>,
 <Type name=__autofs_master>,
 <Type name=__debconf_set_selections>,
 <Type name=__directory>,
 <Type name=__file>,
 <Type name=__group>,
 <Type name=__issue>,
 <Type name=__key_value>,
 <Type name=__link>,
 <Type name=__mkfs>,
 <Type name=__motd>,
 <Type name=__package>,
 <Type name=__package_apt>,
 <Type name=__package_pacman>,
 <Type name=__package_pkg_openbsd>,
 <Type name=__package_rubygem>,
 <Type name=__package_yum>,
 <Type name=__partition_msdos>,
 <Type name=__partition_msdos_apply>,
 <Type name=__postgres_database>,
 <Type name=__postgres_role>,
 <Type name=__process>,
 <Type name=__removeline>,
 <Type name=__ssh_authorized_key>,
 <Type name=__timezone>,
 <Type name=__user>]

In [4]: 

In [4]: oo.Object.
oo.Object.base_dir           oo.Object.changed            oo.Object.list_object_names  oo.Object.list_objects       oo.Object.list_type_names    oo.Object.mro                oo.Object.path

In [4]: oo.Object.list
oo.Object.list_object_names  oo.Object.list_objects       oo.Object.list_type_names    

In [4]: oo.Object.list_ob
oo.Object.list_object_names  oo.Object.list_objects       

In [4]: oo.Object.list_object_names()
Out[4]: <generator object list_object_names at 0x1b759b0>

In [5]: list(oo.Object.list_object_names())
Out[5]: 
['__mkfs/dev/sda1',
 '__mkfs/dev/sda2',
 '__mkfs/dev/sda5',
 '__mkfs/dev/sda6',
 '__mkfs/dev/sda7',
 '__partition_msdos/dev/sda1',
 '__partition_msdos/dev/sda2',
 '__partition_msdos/dev/sda3',
 '__partition_msdos/dev/sda5',
 '__partition_msdos/dev/sda6',
 '__partition_msdos/dev/sda7',
 '__partition_msdos_apply/singleton']

In [6]: list(oo.Object.list_objects())
Out[6]: 
[<Object __mkfs/dev/sda1>,
 <Object __mkfs/dev/sda2>,
 <Object __mkfs/dev/sda5>,
 <Object __mkfs/dev/sda6>,
 <Object __mkfs/dev/sda7>,
 <Object __partition_msdos/dev/sda1>,
 <Object __partition_msdos/dev/sda2>,
 <Object __partition_msdos/dev/sda3>,
 <Object __partition_msdos/dev/sda5>,
 <Object __partition_msdos/dev/sda6>,
 <Object __partition_msdos/dev/sda7>,
 <Object __partition_msdos_apply/singleton>]

In [7]: 


'''

class Type(object):

    @staticmethod
    def base_dir():
        """Return the absolute path to the top level directory where types
        are defined.

        Requires the environment variable '__cdist_base_dir' to be set.

        """
        return os.path.join(
            os.environ['__cdist_base_dir'],
            'conf',
            'type'
        )

    @classmethod
    def list_types(cls):
        """Return a list of type instances"""
        for type_name in cls.list_type_names():
            yield cls(type_name)

    @classmethod
    def list_type_names(cls):
        """Return a list of type names"""
        return os.listdir(cls.base_dir())


    def __init__(self, name):
        self.name = name
        self.__explorers = None
        self.__required_parameters = None
        self.__optional_parameters = None

    def __repr__(self):
        return '<Type name=%s>' % self.name

    @property
    def path(self):
        return os.path.join(
            self.base_dir(),
            self.name
        ) 

    @property
    def is_singleton(self):
        """Check whether a type is a singleton."""
        return os.path.isfile(os.path.join(self.path, "singleton"))

    @property
    def is_install(self):
        """Check whether a type is used for installation (if not: for configuration)"""
        return os.path.isfile(os.path.join(self.path, "install"))

    @property
    def explorers(self):
        """Return a list of available explorers"""
        if not self.__explorers:
            try:
                self.__explorers = os.listdir(os.path.join(self.path, "explorer"))
            except EnvironmentError as e:
                # error ignored
                self.__explorers = []
        return self.__explorers

    @property
    def required_parameters(self):
        """Return a list of required parameters"""
        if not self.__required_parameters:
            parameters = []
            try:
                with open(os.path.join(self.path, "parameter", "required")) as fd:
                    for line in fd:
                        parameters.append(line.strip())
            except EnvironmentError as e:
                # error ignored
                pass
            finally:
                self.__required_parameters = parameters
        return self.__required_parameters

    @property
    def optional_parameters(self):
        """Return a list of optional parameters"""
        if not self.__optional_parameters:
            parameters = []
            try:
                with open(os.path.join(self.path, "parameter", "optional")) as fd:
                    for line in fd:
                        parameters.append(line.strip())
            except EnvironmentError as e:
                # error ignored
                pass
            finally:
                self.__optional_parameters = parameters
        return self.__optional_parameters


class Object(object):

    @staticmethod
    def base_dir():
        """Return the absolute path to the top level directory where objects
        are defined.

        Requires the environment variable '__cdist_out_dir' to be set.

        """
        base_dir = os.path.join(
            os.environ['__cdist_out_dir'],
            'object'
        )
        # FIXME: should directory be created elsewhere?
        if not os.path.isdir(base_dir):
            os.mkdir(base_dir)
        return base_dir

    @classmethod
    def list_objects(cls):
        """Return a list of object instances"""
        for object_name in cls.list_object_names():
            type_name = object_name.split(os.sep)[0]
            object_id = os.sep.join(object_name.split(os.sep)[1:])
            yield cls(Type(type_name), object_id=object_id)

    @classmethod
    def list_type_names(cls):
        """Return a list of type names"""
        return os.listdir(cls.base_dir())

    @classmethod
    def list_object_names(cls):
        """Return a list of object names"""
        for path, dirs, files in os.walk(cls.base_dir()):
            # FIXME: use constant instead of string
            if '.cdist' in dirs:
                yield os.path.relpath(path, cls.base_dir())

    def __init__(self, type, object_id=None, parameter=None, requirements=None):
        self.type = type # instance of Type
        self.object_id = object_id
        self.qualified_name = os.path.join(self.type.name, self.object_id)
        self.parameter = parameter or {}
        self.requirements = requirements or []
        
    def __repr__(self):
        return '<Object %s>' % self.qualified_name

    @property
    def path(self):
        return os.path.join(
            self.base_dir(),
            self.qualified_name,
            '.cdist'
        )

    @property
    def changed(self):
        """Check whether the object has been changed."""
        return os.path.isfile(os.path.join(self.path, "changed"))

    @changed.setter
    def changed(self, value):
        """Change the objects changed status."""
        path = os.path.join(self.path, "changed")
        if value:
            open(path, "w").close()
        else:
            try:
                os.remove(path)
            except EnvironmentError:
                # ignore
                pass

    # FIXME: implement other properties/methods
